torch/csrc/autograd/python_variable.cpp

```diff
@@ -909,9 +903,22 @@ PyTypeObject THPVariableType = {
   0,                                           /* tp_dictoffset */
   nullptr,                                     /* tp_init */
   nullptr,                                     /* tp_alloc */
-  THPVariable_pynew                            /* tp_new */
+  // NB: It is illegal to directly create a _TensorBase.  Instead,
+  // subclass it first (the metaclass will initialize tp_new) and
+  // then construct it
+  nullptr,                                     /* tp_new */
 };
 
+int THPVariableMetaType_init(PyObject *cls, PyObject *args, PyObject *kwargs) {
+  if (PyType_Type.tp_init(cls, args, kwargs) < 0) {
+    return -1;
+  }
+  if (((PyTypeObject*)cls)->tp_base == &THPVariableType) {
```

Comment:
This means that the class being created is a subclass of _TensorBase? If so, it doesn't match the description in the PR where you seem to say that only the _TensorBase itself should have its tp_new modified?

Reply:
I'm not sure what PR description you're referring to. I did get this wrong the first time I wrote the PR, the operant clarification is:

> The subtlety here is that we only install our custom tp_new for direct subclasses of TensorBase. This is important, because overriding the tp_new also overrides any user defined constructor. Fortunately class Tensor(_TensorBase) has no nontrivial constructors and doesn't mind, but other subclasses like Parameter definitely mind!

The logic here is testing for a direct subclass of TensorBase.

Reply:
Ho right, THPVariable is not Tensor, but _TensorBase. So this match the Tensor class!
Sounds good!